src/modelfree/shaping_wrappers.py

```diff
@@ -0,0 +1,208 @@
+
+        # replace rew with differently shaped rew
+        # victim is agent 0, attacker is agent 1
+        for env_num in range(self.num_envs):
+            shaped_reward = 0
+            for rew_type, rew_value in infos[env_num][1].items():
+                if rew_type not in self.shaping_params:
+                    continue
+
+                weighted_reward = self.shaping_params[rew_type] * rew_value
+                self.step_rew_dict[rew_type][env_num].append(weighted_reward)
+                if self.reward_annealer is not None:
+                    c = self.get_annealed_exploration_reward()
+                    if rew_type == 'reward_remaining':
+                        weighted_reward *= (1 - c)
+                    else:
+                        weighted_reward *= c
+                shaped_reward += weighted_reward
+
+            if done[env_num]:
+                for rew_type in self.step_rew_dict:
+                    self.ep_rew_dict[rew_type].append(sum(self.step_rew_dict[rew_type][env_num]))
+                    self.step_rew_dict[rew_type][env_num] = []
+            self.counter += 1
+            self._log_sparse_dense_rewards()
+            rew[env_num] = shaped_reward
+
+        return obs, rew, done, infos
+
+    def get_annealed_exploration_reward(self):
+        """
+        Returns c (which will be annealed to zero) s.t. the total reward equals
+        c * reward_move + (1 - c) * reward_remaining
+        """
+        c = self.reward_annealer()
+        assert 0 <= c <= 1
+        return c
+
+
+class NoisyAgentWrapper(BaseRLModel):
+    def __init__(self, agent, noise_annealer, noise_type='gaussian'):
+        """
+        Wrap an agent and add noise to its actions
+        :param agent: PolicyToModel(BaseRLModel) the agent to wrap
+        :param noise_annealer: Annealer.get_value - presumably the noise should be decreased
+        over time in order to get the adversarial policy to perform well on a normal victim.
+        :param noise_type: str - the type of noise parametrized by noise_annealer's value.
+        Current options are [gaussian]
+        """
+        self.agent = agent
+        self.sess = agent.sess
+        self.noise_annealer = noise_annealer
+        self.noise_generator = self._get_noise_generator(noise_type)
+
+    def _get_noise_generator(self, noise_type):
+        noise_generators = {
+            'gaussian': lambda x, size: np.random.normal(scale=x, size=size)
+        }
+        return noise_generators[noise_type]
+
+    def predict(self, observation, state=None, mask=None, deterministic=False):
+        original_actions, states = self.agent.predict(observation, state, mask, deterministic)
+        action_shape = original_actions.shape
+
+        noise_param = self.noise_annealer()
+        noise = self.noise_generator(noise_param, action_shape)
+        normed_noise = original_actions * noise
+
+        noisy_actions = original_actions + normed_noise
```

Comment:
Or `original_actions * (1 + noise)`?

It's not clear to me whether additive or multiplicative noise will work best; let's discuss in person.

Reply:
Yes that is clearer.